src/org/exist/collections/Collection.java

```diff
@@ -1096,6 +1096,20 @@ public void removeCollection(final DBBroker broker, final XmldbURI name) throws
         }
     }
 
+    public void removeResource(Txn txn, DBBroker broker, DocumentImpl doc) throws PermissionDeniedException, LockException, IOException, TriggerException {
+        if (doc == null) return;
+
+        if (doc.getCollection() != this) {
```

Comment:
as far as DocumentImpl allowed outside eXist that must stay here because document can be moved concurrently between get document object and call `removeResource`. And why you don't want to help to detect such bad situation?